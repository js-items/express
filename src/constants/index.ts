export const TEST_URL = '/items'