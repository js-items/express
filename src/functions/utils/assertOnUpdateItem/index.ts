// tslint:disable:no-any
import { UpdateItem } from '@js-items/foundation';
import { TestItem } from '@js-items/foundation/dist/functions/utils/testItem';
import { OK } from 'http-status-codes';
import supertest from 'supertest';
import createTestUrl from '../createTestUrl';

const assertOnUpdateItem = async <I extends TestItem>({
  client,
  status = OK,
  params,
  body,
  expectedParams,
  url,
  updateItem,
}: Options<I>) => {
  const requestUrl =
    params !== undefined ? createTestUrl({ url, params }) : url;

  const result = await client.patch(requestUrl).send(body);

  expect(result.status).toBe(status);
  expect(result.body).toMatchSnapshot();

  if (expectedParams !== undefined) {
    expect(updateItem).toBeCalledWith(expectedParams);
  }
};

export interface Options<I extends TestItem> {
  readonly client: supertest.SuperTest<supertest.Test>;
  readonly status?: number;
  readonly params?: any;
  readonly body?: any;
  readonly expectedParams?: any;
  readonly url: string;
  readonly updateItem: UpdateItem<I>;
}

export default assertOnUpdateItem;
