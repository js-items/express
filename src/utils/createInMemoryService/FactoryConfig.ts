import {
  CountItems,
  CreateItem,
  DeleteItem,
  DeleteItems,
  GetItem,
  GetItems,
  Item,
  ReplaceItem,
  UpdateItem,
} from '@js-items/foundation';

export default interface FactoryConfig<I extends Item> {
  readonly countItems?: CountItems<I>;
  readonly createItem?: CreateItem<I>;
  readonly updateItem?: UpdateItem<I>;
  readonly replaceItem?: ReplaceItem<I>;
  readonly deleteItem?: DeleteItem<I>;
  readonly deleteItems?: DeleteItems<I>;
  readonly getItem?: GetItem<I>;
  readonly getItems?: GetItems<I>;
}
