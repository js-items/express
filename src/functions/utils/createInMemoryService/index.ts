// tslint:disable:no-object-literal-type-assertion
import { Item } from '@js-items/foundation';
import Facade from '@js-items/foundation/dist/Facade';
import {
  Options as CreateItemOptions,
  Result as CreateItemResult,
} from '@js-items/foundation/dist/functions/CreateItem';
import {
  Options as GetItemOptions,
  Result as GetItemResult,
} from '@js-items/foundation/dist/functions/GetItem';
import { Result as GetItemsResult } from '@js-items/foundation/dist/functions/GetItems';
import { Result as ReplaceItemResult } from '@js-items/foundation/dist/functions/ReplaceItem';
import { Result as UpdateItemResult } from '@js-items/foundation/dist/functions/UpdateItem';
import { start } from '@js-items/foundation/dist/interfaces/Cursor';
import FactoryConfig from './FactoryConfig';

const defaultPromise = () => Promise.resolve();
/* istanbul ignore next */
const createInMemoryService = <I extends Item>(
  factoryConfig: FactoryConfig<I>
): Facade<I> => ({
  countItems: () => Promise.resolve({ count: 1 }),
  createItem: ({ item }: CreateItemOptions<I>) =>
    Promise.resolve({ item } as CreateItemResult<I>),
  deleteItem: defaultPromise,
  deleteItems: defaultPromise,
  getItem: ({ id }: GetItemOptions<I>) =>
    Promise.resolve({ item: { id } } as GetItemResult<I>),
  getItems: () =>
    Promise.resolve({
      cursor: {
        after: start,
        before: start,
        hasAfter: false,
        hasBefore: false,
      },
      items: [{ id: '1' }, { id: '2' }],
    } as GetItemsResult<I>),
  replaceItem: () =>
    Promise.resolve({ item: { id: '2' } } as ReplaceItemResult<I>),
  updateItem: () =>
    Promise.resolve({ item: { id: '1' } } as UpdateItemResult<I>),
  ...factoryConfig,
});

export default createInMemoryService;
