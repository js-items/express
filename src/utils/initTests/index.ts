import * as sourceMapSupport from 'source-map-support';
import { TEST_URL } from '../../constants';
sourceMapSupport.install();
import dotenv from 'dotenv';
dotenv.config();
import { Item } from '@js-items/foundation';
import express from 'express';
import createSupertest from 'supertest';
import itemsFactory from '../../factory';
import FactoryConfig from '../../FactoryConfig';

export default <I extends Item>(factoryOptions: FactoryConfig<I>) => {
  const app: express.Application = express();

  app.use(TEST_URL, itemsFactory(factoryOptions));

  const request = createSupertest(app);

  return { request };
};
