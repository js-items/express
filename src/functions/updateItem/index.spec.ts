// tslint:disable:no-any
import * as sourceMapSupport from 'source-map-support';
import assertOnUpdateItem from '../utils/assertOnUpdateItem';
import initTests from '../utils/initTests';
sourceMapSupport.install();
import dotenv from 'dotenv';
dotenv.config();
import { ItemNotFoundError } from '@js-items/foundation';
import {
  Options as UpdateItemOptions,
  Result as UpdateItemResult,
} from '@js-items/foundation/dist/functions/UpdateItem';
import testItem, {
  TestItem,
} from '@js-items/foundation/dist/functions/utils/testItem';
import { NOT_FOUND, OK } from 'http-status-codes';
import { TEST_URL } from '../../constants';
import createInMemoryService from '../utils/createInMemoryService';

jest.mock('uuid', () => ({
  v4: jest.fn(() => '1'),
}));

const notExistingId = 'not-existing-id';

const updateItem = jest.fn(
  async ({ patch, id }: UpdateItemOptions<TestItem>) => {
    if (id !== testItem.id) {
      return Promise.reject(new ItemNotFoundError('item', testItem.id));
    }

    // tslint:disable-next-line:no-object-literal-type-assertion
    return Promise.resolve({ item: patch } as UpdateItemResult<TestItem>);
  }
);

describe('@updateItem', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  const service = createInMemoryService({
    updateItem,
  });

  const { request } = initTests<TestItem>({ service });

  const expectedParams = {
    filter: {},
    id: testItem.id,
    patch: testItem,
  };

  const defaultOptions = {
    body: testItem,
    client: request,
    expectedParams,
    updateItem,
    url: `${TEST_URL}/${testItem.id}`,
  };

  const notFoundOptions = {
    body: { ...testItem, id: notExistingId },
    expectedParams: {
      ...expectedParams,
      id: notExistingId,
      patch: { ...testItem, id: notExistingId },
    },
  };

  it('throws not found error when item does exist', async () => {
    await assertOnUpdateItem({
      ...defaultOptions,
      ...notFoundOptions,
      status: NOT_FOUND,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('throws not found error when item does exist and envelope is enabled', async () => {
    await assertOnUpdateItem({
      ...defaultOptions,
      ...notFoundOptions,
      params: {
        envelope: true,
      },
      status: OK,
      url: `${TEST_URL}/${notExistingId}`,
    });
  });

  it('updates item', async () => {
    const updatedItem = {
      booleanProperty: true,
      id: testItem.id,
      numberProperty: 999,
      stringProperty: 'testowo',
    };

    await assertOnUpdateItem({
      ...defaultOptions,
      body: updatedItem,
      expectedParams: {
        ...expectedParams,
        patch: updatedItem,
      },
    });
  });
  // tslint:disable-next-line:max-file-line-count
});
