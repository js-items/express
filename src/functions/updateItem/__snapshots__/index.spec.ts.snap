// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@updateItem throws not found error when item does exist 1`] = `
Object {
  "itemId": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
  "itemName": "item",
  "transactionId": "1",
}
`;

exports[`@updateItem throws not found error when item does exist and envelope is enabled 1`] = `
Object {
  "headers": Object {
    "x-powered-by": "Express",
  },
  "response": Object {
    "itemId": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
    "itemName": "item",
    "transactionId": "1",
  },
  "status": 404,
}
`;

exports[`@updateItem updates item 1`] = `
Object {
  "booleanProperty": true,
  "id": "6da9dce6-be40-4fac-bd9c-6a2f348a1c87",
  "numberProperty": 999,
  "stringProperty": "testowo",
}
`;
